### PR TITLE
[STEP 1.3] Update query and pagination

### PR DESCRIPTION
**Pull request by AndreHGN**
## What does this PR do?
First the GraphQL query has been updated to pick up the amount of pages of the response. Mainly this PR allows the user to check on other episodes through pagination. Three types of buttons were added to control the pagination: 

1. Next button: go to next page, if it exists;
2. Previous button: go to previous page, if it exists;
3. Numbered button: go to the page corresponding to the number;

Furthermore, it indicates which page is being shown highlighting one of the numbered button.

### CODE REVIEW

File: src/getEpisodes.js [hunk 1]
```diff
@@ -1,31 +1,47 @@
...
+            }
+        });
+        return response.data.data.episodes;
+    }
+    catch (error) {
+        printError(error.message);
+    }
+    finally {
+        removeLoader();
+    }
```
Comment (lucas-eugenio) [inline src/getEpisodes.js]: getEpisodes has the responsibility to get the data you need from the server. I don think it makes much sense to make it also responsible of adding error message and removing loading.

On the previous version, the only functions that created visual elements were "printError" and "renderData". Now, you are breaking the patter that you created.

If you really want to, at least give it a name that embraces all this responsibilities.

File: src/index.js [hunk 2]
```diff
@@ -1,25 +1,19 @@
...
-        const episodes = response.data.data.episodes;
-        renderData(episodes);
-    }
-    catch(error) {
-        printError(error.message);
-    }
-    finally{
-        removeLoader();
-    }
+async function initialize() {
```
Comment (lucas-eugenio) [inline src/index.js:8]: I really like this name!

File: src/makeCard.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 export default makeCard;
 
 function isEven(number) {
-    return number % 2 === 0 ? true : false; 
+    return number % 2 === 0;
```
Comment (lucas-eugenio) [inline src/makeCard.js:4]: :+1:

File: src/makePageButtons.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+    pageButton.innerHTML = `${pageNumber}`;
+
+    return pageButton;
+
+}
+
+function makePageButtons(currentPage, maxPage) {
+
+    const pageButtons = [...Array(maxPage)].map((_, i) => makeButton(i+1));
```
Comment (lucas-eugenio) [inline src/makePageButtons.js]: Why are you creating this Array to deconstruct? Recommend the cleaner version:
```
const pageButtons = Array(maxPage).fill().map((_, i) => makeButton(i+1));
``` 

Also, pageButtons let us think that all buttons of the page is defined by this. Here, we have only the numeric buttons, you should call this *pageNumericButtons*

File: src/renderElements.js [hunk 5]
```diff
@@ -29,7 +28,21 @@ export function renderLoader() {
...
 
-export function removeLoader() {
-    const loader = document.getElementsByClassName("loader").item(0);
-    loader.remove();
-}
\ No newline at end of file
+export function renderPagination(currentPage, maxPage) {
+
+    const pageButtonsTop = makePageButtons(currentPage, maxPage);
+    const pageButtonsBottom = makePageButtons(currentPage, maxPage);
```
Comment (lucas-eugenio) [inline src/renderElements.js]: If they are the same thing, why should we treat them as separate things?

File: src/renderElements.js [hunk 6]
```diff
@@ -29,7 +28,21 @@ export function renderLoader() {
...
+    const paginationDivs = document.getElementsByClassName("pagination");
+    const nextPageButtons = document.getElementsByClassName("next");
+
+
+    pageButtonsTop.forEach(button => {
+        paginationDivs.item(0).insertBefore(button, nextPageButtons.item(0));
+    });
+
+    pageButtonsBottom.forEach(button => {
+        paginationDivs.item(1).insertBefore(button, nextPageButtons.item(1));
```
Comment (lucas-eugenio) [inline src/renderElements.js]: Same here

File: src/setupPagination.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import updateData from "./updateData.js";
+
+export default setupPagination;
+
+function updateSelectedButton(currentPage) {
+    const oldButtons = document.getElementsByClassName("selected");
```
Comment (lucas-eugenio) [inline src/setupPagination.js:6]: Why buttons instead of button?

File: src/setupPagination.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+function goToNextPage(currentPage, maxPage) {
+    return function() {
+        if (currentPage !== maxPage) {
+            currentPage += 1;
+            updateData(currentPage);
+            updateSelectedButton(currentPage);
+        }
+    }
+}
```
Comment (lucas-eugenio) [inline src/setupPagination.js]: Why are you returning a function here? I don't get it.

**Review by lucas-eugenio — CHANGES_REQUESTED** [review]
There are two UX behaviors that I think you must add.

- Loading State on Pages Count. We already talked about it on previous version, you must show to the user that something is happening.
- Add disabled behavior on buttons when you cannot click them. What you think that the user might think if he clicks a button and nothing happens?

Do not hide things from the user. If you are requesting data, show this to the user. If he/she can not click on something, show this. You can not assume that he knows how your site works.

File: src/renderElements.js (hunk 5 above)
Comment (AndreHGN) [reply src/renderElements.js]: My intent here was to render the same numeric buttons on two different parts of the webpage. One inside the top pagination div and the other one inside the bottom pagination div. The first thing I tried was this:

```
const pageNumericButtons = makePageButtons(currentPage, maxPage);

pageNumericButtons.forEach(button => {
        paginationDivs.item(0).insertBefore(button, nextPageButtons.item(0));
        paginationDivs.item(1).insertBefore(button, nextPageButtons.item(1));
    });
```

But the problem is that for some reason, this method only inserts the buttons on the last paginationDiv described. The code above results in the following structure:

image

If we change the order of the paginationDivs, it also changes where the buttons are inserted:
```
pageNumericButtons.forEach(button => {
        paginationDivs.item(1).insertBefore(button, nextPageButtons.item(1));
        paginationDivs.item(0).insertBefore(button, nextPageButtons.item(0));
    });
```
Resulting:

image

File: src/setupPagination.js (hunk 7 above)
Comment (AndreHGN) [reply src/setupPagination.js:6]: Considering there are two sets of the same buttons (one at the top of the page and the other at the bottom) there would also be in total two numeric buttons highlighted (selected).

So I though to put the name of the parameter in plural to reference these two buttons.

I guess following that logic I should have named the function in plural too (updateSelectedButtons). That's my bad..

File: src/renderElements.js (hunk 5 above)
Comment (AndreHGN) [reply src/renderElements.js]: Oh I think I found a solution that suits better than what I did previously. Using the method `.cloneNode` solves the problem:
```
const pageNumericButtons = makePageButtons(currentPage, maxPage);

pageNumericButtons.forEach(button => {
        paginationDivs.item(0).insertBefore(button, nextPageButtons.item(0));
        paginationDivs.item(1).insertBefore(button.cloneNode(true), nextPageButtons.item(1));
    });
```

File: src/setupPagination.js (hunk 8 above)
Comment (AndreHGN) [reply src/setupPagination.js]: I was having a problem with the `.onclick` method. The functions that were passed to it were automatically being called.

It is the same problem as the one from this post: https://stackoverflow.com/questions/16310423/addeventlistener-calls-the-function-without-me-even-asking-it-to

I applied the answer that seemed the cleaner to read. But in our pair up meeting Rafael told me a better way that uses arrow function to solve this.

**Review by lucas-eugenio — APPROVED** [review]
LGTM!

**Pull request merged**